clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java

```diff
@@ -522,48 +579,103 @@ public void onMemberEpochUpdated(Optional<Integer> memberEpochOpt, Optional<Stri
      * Represents a request to acknowledge delivery that can be retried or aborted.
      * ** UNDER CONSTRUCTION **
      */
-    static class AcknowledgeRequestState extends RequestState {
+    class AcknowledgeRequestState extends RequestState {
 
         /**
          * The node to send the request to.
          */
-        private final int node;
+        private final int nodeId;
 
         /**
          * The map of acknowledgements to send
          */
         private final Map<TopicIdPartition, Acknowledgements> acknowledgementsMap;
 
-        /**
-         * Future with the result of the request.
-         */
-        private final CompletableFuture<Void> future;
-
         /**
          * Time until which the request should be retried if it fails with retriable
          * errors. If not present, the request is triggered without waiting for a response or
          * retrying.
          */
         private final Optional<Long> expirationTimeMs;
+        private boolean isExpired = false;
 
         AcknowledgeRequestState(LogContext logContext, String owner,
                                 long retryBackoffMs, long retryBackoffMaxMs,
                                 Optional<Long> expirationTimeMs,
-                                int node,
+                                int nodeId,
                                 Map<TopicIdPartition, Acknowledgements> acknowledgementsMap) {
             super(logContext, owner, retryBackoffMs, retryBackoffMaxMs);
             this.expirationTimeMs = expirationTimeMs;
-            this.node = node;
+            this.nodeId = nodeId;
             this.acknowledgementsMap = acknowledgementsMap;
-            this.future = new CompletableFuture<>();
         }
 
-        UnsentRequest buildRequest() {
-            return null;
+        UnsentRequest buildRequest(long currentTimeMs) {
```

Comment:
Apart from the response handlers, these two methods look the same to me. They should be one method with the handlers as parameters.